### ISSUE TITLE
[API Proposal]: New method HashSet<T>.AddOrUpdate that replaces an existing element if found

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#101003]
### Background and motivation

When adding an existing key to a `Dictionary<TKey,TValue>` collection, there is an option to either preserve or replace the existing value.

```csharp
dictionary.TryAdd(key, value); // Preserve
dictionary[key] = value; // Replace
```

There is no such option for the [`HashSet<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1) collection. There is only an `Add` method with preserving behavior.
So this proposal is to introduce a new `AddOrUpdate` method with replacing behavior:

### API Proposal

```csharp
namespace System.Collections.Generic;

public class HashSet<T>
{
    /// <summary>Adds the specified element to a set, replacing an existing equal element if found.</summary>
    /// <param name="item">The element to add to the set.</param>
    public void AddOrUpdate(T item);
}
```

### API Usage

Recently I wrote a custom `ICollection<T>` collection intended for JSON serialization, that has the functionality of the [`KeyedCollection<TKey,TItem>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.keyedcollection-2), but instead of a `List`+`Dictionary` combo is backed by a `HashSet<T>`. Practically it is a dictionary that is serialized in JSON as a list. I wanted to replicate the existing deserialization behavior of the `Dictionary<TKey,TValue>` collection regarding duplicate keys, which is to preserve the last key found in the JSON document and silently discard the earlier duplicates. So I implemented the `ICollection<TItem>.Add` method like this:

```csharp
abstract class JsonKeyedCollectionBase<TKey, TItem> : ICollection<TItem>
{
    private readonly HashSet<TItem> _set;

    /* ... */

    void ICollection<TItem>.Add(TItem item)
    {
        if (_set.Add(item)) return;
        bool removed = _set.Remove(item);
        bool added = _set.Add(item);
        Debug.Assert(removed);
        Debug.Assert(added);
    }
}
```

The whole implementation can be found [here](https://gist.github.com/theodorzoulias/46936d9c04fb45630a0d6782d6ce2546). If the proposed API was available, I could make the above implementation simpler and (presumably) more efficient like this:

```csharp
    void ICollection<TItem>.Add(TItem item) => _set.AddOrUpdate(item);
```

**Comment by hez2010** [comment]
Maybe a new method called `AddOrUpdate` is better?

**Comment by theodorzoulias** [comment]
@hez2010 do you mean the same method signature with a different name, or a different signature? In the second case, could you share the signature of the `AddOrUpdate` method?

**Comment by hez2010** [comment]
```cs
void AddOrUpdate(T value, Func<T, T> updateValueFactory);
```

Usage:
```cs
hashSet.AddOrUpdate(42, oldValue => 43);
```

This also matches the existing `ConcurrentDictionary.AddOrUpdate` API.

**Comment by hez2010** [comment]
Yeah I mean the same method signature with a different name. 
We already have `ConcurrentDictionary.AddOrUpdate` so this can be consistent with how we name things in BCL.

**Comment by theodorzoulias** [comment]
@hez2010 using this `AddOrUpdate` in my specific scenario, I would have to do:

```
void ICollection<TItem>.Add(TItem item) => _set.Add(item, _ => item);
```

...which would capture the `item` argument, resulting in an allocation of a closure object. I would prefer to avoid that.

Also could you share a scenario where invoking the `updateValueFactory` and returning a value different than the `item`, would make sense? The `HashSet<T>` is not typically used as a dictionary, and my specific scenario already deviates from the typical usage of this collection.

Regarding the other `AddOrUpdate` which returns a `bool`, it is essentially the same with my `Add` proposal, changing only how the `replaced` information is returned. I don't have a strong opinion about this. Both a return value and an `out` parameter are fine with me.

I wouldn't even mind if this information was omitted. In my specific scenario I don't care if there was already an equal item in the collection, neither I care what was this item.

**Comment by hez2010** [comment]
Sorry for the misleading. I found the updateValueFactory is problematic here so I deleted the comment just after I posted it.

**Comment by theodorzoulias** [comment]
@hez2010 no problem. I edited my previous comment accordingly.

**Comment by DaZombieKiller** [comment]
If we had `CollectionsMarshal` APIs for `HashSet<T>` like we do for `Dictionary<TKey, TValue>` it would be possible to implement this manually without multiple lookups:
```cs
public static partial class CollectionsMarshal
{
    public static ref T GetValueRefOrAdd<T>(HashSet<T> set, T value, out bool exists);
}

// ...

public static void AddOrReplace<T>(this HashSet<T> set, T value)
{
    CollectionsMarshal.GetValueRefOrAdd(set, value, out _) = value;
}
```

**Comment by stephentoub** [comment]
> If we had CollectionsMarshal APIs for HashSet<T> like we do for Dictionary<TKey, TValue>

There's a reason, though, that those APIs for dictionary don't give you a ref to the TKey: it'd be incredibly easy to accidentally change the key to once that hashes differently, at which point the dictionary would be corrupted. For a HashSet, the value is the key, so providing a ref to it incurs the same problems. It's why we haven't exposed such an API when it's been discussed before.

**Comment by theodorzoulias** [comment]
@eiriktsarpalis I chose the `HashSet<T>` over a `Dictionary<K,V>` because I want to avoid wasting memory by storing the key in two places (the same that @magol wanted [here](https://github.com/dotnet/runtime/issues/18376 "Proposal: Dictionary with key and value in same object")). My collections are stored in memory from the moment my program begins until it terminates, and the program already uses quite a lot of memory, so I want to minimize any wasteful usage of it.

The issue [#21603](https://github.com/dotnet/runtime/issues/21603 "HashSet.GetOrAdd(T)") is about a `HashSet<T>.GetOrAdd` method, that either adds a value or returns the existing equal value. That functionality is distinct from what is asked here, which is a method that always adds a value, regardless of whether an existing equal value exists.

I can see that this proposal could be rejected for the same reason that the other was rejected, because each new API in commonly used collections increases the static size of applications. My two cents are: AFAIK the `HashSet<T>` is not a commonly used collection, and it cannot be amended with a [`CollectionsMarshal`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal) API that would allow the developers to implement efficiently any additional APIs they need in a case-by-case basis (as Stephen Toub explained in a previous comment).

Personally I think that both the `GetOrAdd` and the replacing `Add` are fundamental operations, and in an ideal world they would be part of the public API of the `HashSet<T>` collection.

**Comment by theodorzoulias** [comment]
@eiriktsarpalis btw would you consider reopening the issue [#21603](https://github.com/dotnet/runtime/issues/21603)? Developers can't comment or provide any feedback on closed issues, and the existence of a closed issue discourages developers from opening a new issue with the same proposal. The issue #21603 was closed partially on the anticipation of a future `CollectionsMarshal` API that would enable efficient `GetOrAdd` implementations, and we now know that such an API is not viable for the `HashSet<T>` collection.

**Comment by theodorzoulias** [comment]
I changed the name of the proposed API from `Add` to `AddOrUpdate`, which seems more popular. I also simplified it by removing the `out bool replaced` and `out T oldItem` parameters, because I can't think of a scenario where they would be useful. If such a scenario exists, they could be added as a method overload in the future.

**Comment by eiriktsarpalis** [comment]
> I chose the `HashSet<T>` over a `Dictionary<K,V>` because I want to avoid wasting memory by storing the key in two places

Can you share any benchmarks demonstrating measurable performance issues using dictionary over hashset?

> btw would you consider reopening the issue https://github.com/dotnet/runtime/issues/21603?

I've unlocked it but didn't reopen it.

**Comment by theodorzoulias** [comment]
Sure. Starting with a struct of size 32 bytes that contains a `string` key, storing this struct in a `Dictionary<K,V>` allocates
about 20% more memory compared to storing it in a `HashSet<T>`. The code that produces this statistic is [here](https://dotnetfiddle.net/zssiXN), with this output:

```none
Record size: 32 bytes
Records to store: 100,000
HashSet allocated 4,781,448 bytes, Capacity: 108,631
Dictionary allocated 5,649,112 bytes, Capacity: 108,631
```

Obviously the percentile difference decreases as the size of the struct increases.

> I've unlocked it but didn't reopen it.

Won't it get locked again automatically after a few weeks of inactivity? The idea is to allow developers who find this feature useful, to support it by upvoting it.

**Comment by eiriktsarpalis** [comment]
I think what I find problematic with this proposal is that it presupposes two distinct definitions of equality on the element type: one governing lookup semantics and another being the effective semantics (reference equality or something else). It's effectively encoding a dictionary but in a way that sounds _very_ error prone.

I would personally just use a dictionary in this case. Assuming the overhead of storing values is causing performance issues in an applications, it sounds like a case where using a custom collection might be warranted.

**Comment by theodorzoulias** [comment]
@eiriktsarpalis in my application I have encapsulated the equality definition for lookups in a dedicated collection class ([`JsonKeyedCollectionBase<TKey, TItem>`](https://gist.github.com/theodorzoulias/46936d9c04fb45630a0d6782d6ce2546)), which I have tested and verified that works correctly. I would prefer if a [`KeyedDictionary<TKey, TValue>`](https://github.com/dotnet/runtime/issues/18376) was natively available, because the conversion `TKey` -> `TItem` -> `TKey` harms the performance of lookups, but since this doesn't exist I am OK with what I get from the humble `HashSet<T>`. For my specific use case, memory efficiency is more important than performance.
